src/wbase/UserQueryInfo.h

```diff
@@ -0,0 +1,123 @@
+// -*- LSST-C++ -*-
+
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_WBASE_USERQUERYINFO_H
+#define LSST_QSERV_WBASE_USERQUERYINFO_H
+
+// System headers
+#include <map>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+// Qserv headers
+#include "global/intTypes.h"
+
+
+// This header declarations
+namespace lsst::qserv::wbase {
+
+/// &&& doc
+class UserQueryInfo {
+public:
+    using Ptr = std::shared_ptr<UserQueryInfo>;
+    using Map = std::map<QueryId, UserQueryInfo::Ptr>;
+
+
+    static Ptr uqMapInsert(QueryId qId, std::vector<std::string> const& templateStr);
+    static Ptr uqMapGet(QueryId);
+
+    UserQueryInfo(QueryId);
+    UserQueryInfo() = delete;
+    UserQueryInfo(UserQueryInfo const&) = delete;
+    UserQueryInfo& operator=(UserQueryInfo const&) = delete;
+
+    ~UserQueryInfo() = default;
+
+    /// &&& doc
+    size_t addTemplate(std::string const& templateStr);
```

Comment:
_uqMtx was supposed to be used to protect that. Good catch, I'm not sure what happened there.

Reply:
Making it public makes it a lot easier to add new templates.

Reply:
My main concern was that `_templates` can be populated in 2 ways - first by calling `addTemplate` which is actually fine to me because the method would extend the collection and return a stable (provided you're not going to introduce operations for deleting elements) index of the added templated within the collection. However, you also have an alternative way of populating `_templates`  via the static method `uqMapInsert` where you're sending in a collection of templates for which you don't have any clear semantic of "template id". I find this dangerous and confusing. Do you really want  that`uqMapInsert` to be present in the API?

Reply:
Now I am confused. The only way of adding templates to `_templates` is to call `addTemplate`.

`uqMapInsert` adds a new `UserQueryInfo` object to the static `_uqMap` with key `QueryId`. Each `UserQueryInfo` instance has its own `_templates` vector, which needs to be filled by calling `addTemplate`.